### ISSUE TITLE
Relayer is not properly getting the lastest app state when executing handshake steps

### ISSUE THREAD

**Issue by colin-axner** [cosmos/relayer#300]
```
rly tx link cosmicslippers -d -o 3s 
I[2020-10-21|10:25:39.820] - [slippy-0] -> creating client (thestargate) for [stargate-4]header-height{91492} trust-period(336h0m0s) 
I[2020-10-21|10:25:44.224] ✔ [slippy-0]@{99288} - msg(0:create_client) hash(ED7014185BFE2810669B5AECC21AE8E569B8381787DCDB836280FEE0A012F576) 
I[2020-10-21|10:25:44.224] ★ Clients created: [slippy-0]client(thestargate) and [stargate-4]client(slipperland) 
I[2020-10-21|10:25:45.738] - [slippy-0]@{0}conn(cosmicslippers)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{0}conn(theslipperportal)-{STATE_UNINITIALIZED_UNSPECIFIED} 
I[2020-10-21|10:25:49.226] ✔ [slippy-0]@{99289} - msg(0:connection_open_init) hash(A66AF9EC485F1A6AF2C67D6DFB8F545860B05DF9462CD91718FA5F8E35837938) 
I[2020-10-21|10:25:49.784] - [slippy-0]@{0}conn(cosmicslippers)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{0}conn(theslipperportal)-{STATE_UNINITIALIZED_UNSPECIFIED} 
I[2020-10-21|10:25:49.792] - [slippy-0] -> sending transaction:         
E[2020-10-21|10:25:49.793] - [slippy-0] -> err(failed to execute message; message index: 0: connection handshake open init failed: cosmicslippers: connection already exists: invalid request) 
^C
```

When executing the connection handshake, the first attempt at INIT on the source chain succeeds as show above by the first set of steps:
```
rly tx link cosmicslippers -d -o 3s 
I[2020-10-21|10:25:39.820] - [slippy-0] -> creating client (thestargate) for [stargate-4]header-height{91492} trust-period(336h0m0s) 
I[2020-10-21|10:25:44.224] ✔ [slippy-0]@{99288} - msg(0:create_client) hash(ED7014185BFE2810669B5AECC21AE8E569B8381787DCDB836280FEE0A012F576) 
I[2020-10-21|10:25:44.224] ★ Clients created: [slippy-0]client(thestargate) and [stargate-4]client(slipperland) 
I[2020-10-21|10:25:45.738] - [slippy-0]@{0}conn(cosmicslippers)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{0}conn(theslipperportal)-{STATE_UNINITIALIZED_UNSPECIFIED} 
I[2020-10-21|10:25:49.226] ✔ [slippy-0]@{99289} - msg(0:connection_open_init) hash(A66AF9EC485F1A6AF2C67D6DFB8F545860B05DF9462CD91718FA5F8E35837938) 
```

The first iteration of the for loop ends [here](https://github.com/ovrclk/relayer/blob/06740742e8e88baaff21efa5633c81a170d5ea24/relayer/connection-tx.go#L55)

Using the debug logs we can deduce that the second iteration of the for loop is executed in the exact same way as the first iteration. 

The next debug log shows:
```
I[2020-10-21|10:25:49.784] - [slippy-0]@{0}conn(cosmicslippers)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{0}conn(theslipperportal)-{STATE_UNINITIALIZED_UNSPECIFIED} 
```

This tells us both source and destination have uninitialized states. This is contradictory to the successful connection open init that occurred just before. The connection states [queried](https://github.com/ovrclk/relayer/blob/06740742e8e88baaff21efa5633c81a170d5ea24/relayer/connection-tx.go#L111) must be outdated resulting in execution of the same message twice. This gives us the expected result of:
```
E[2020-10-21|10:25:49.793] - [slippy-0] -> err(failed to execute message; message index: 0: connection handshake open init failed: cosmicslippers: connection already exists: invalid request) 
```

I can confirm the connection does exist
```
gaiad q ibc connection connections
connections:
- client_id: thestargate
  counterparty:
    client_id: slipperland
    connection_id: theslipperportal
    prefix:
      key_prefix: aWJj
  id: cosmicslippers
  state: STATE_INIT
  versions:
  - "\n\x011\x12\rORDER_ORDERED\x12\x0FORDER_UNORDERED"
height:
  version_height: "99345"
  version_number: "0"
pagination:
  next_key: null
  total: "0"
```

We need to ensure the light clients are updated to the latest possible states

**Comment by cwgoes** [comment]
The relayer ought to wait for the transaction to be confirmed before it moves on with the handshake.

Is this not happening somehow? Or are the height offsets wrong?

**Comment by colin-axner** [comment]
Did some digging on this. The relayer waits for the transaction to be committed, but queries from the header - 1 height causing the connection not to be updated at this height. Removing all the `-1`s leads to new issues.

I'm not sure if this is also true, but it is possible that txs committed at height h will not be in effect until queries at height h + 1 meaning the relayer would need to wait for an additional block after it is committed

**Comment by AdityaSripal** [comment]
This issue exists for the entire handshake process:

```
➜  relayer git:(aditya/testing) ✗ rly tx full-path demo -d -o 3s
I[2020-10-21|16:29:16.003] - [slippy-0]@{0}conn(adityadstconn)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{94536}conn(adityasrcconn)-{STATE_INIT} 
I[2020-10-21|16:29:16.305] - [slippy-0] -> sending transaction:         
E[2020-10-21|16:29:16.305] - [slippy-0] -> err(failed to execute message; message index: 1: connection handshake open try failed: failed connection state verification for client (adityadstclient): consensus state does not exist for height {0 94536}: consensus state not found: invalid request) 
^C
➜  relayer git:(aditya/testing) ✗ make install
installing rly binary...
➜  relayer git:(aditya/testing) ✗ rly tx full-path demo -d -o 3s
I[2020-10-21|16:33:04.763] - [slippy-0]@{0}conn(adityadstconn)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{94574}conn(adityasrcconn)-{STATE_INIT} 
I[2020-10-21|16:33:09.225] ✔ [slippy-0]@{102969} - msg(0:update_client,1:connection_open_try) hash(386D3B65BB0DA6D95490BE920C98E96CD79AF09BB8D6F161FB03C54FDAFDA7B9) 
I[2020-10-21|16:33:10.766] - [slippy-0]@{0}conn(adityadstconn)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{94575}conn(adityasrcconn)-{STATE_INIT} 
I[2020-10-21|16:33:11.119] - [slippy-0] -> sending transaction:         
E[2020-10-21|16:33:11.119] - [slippy-0] -> err(failed to execute message; message index: 1: connection handshake open try failed: cannot relay connection attempt: invalid connection: invalid request) 
^C
➜  relayer git:(aditya/testing) ✗ rly tx full-path demo -d -o 3s
I[2020-10-21|16:33:36.831] - [stargate-4]@{94580}conn(adityasrcconn)-{STATE_INIT} : [slippy-0]@{102974}conn(adityadstconn)-{STATE_TRYOPEN} 
I[2020-10-21|16:33:39.622] ✔ [stargate-4]@{94581} - msg(0:update_client,1:connection_open_ack) hash(F8F7234D3FD4E77EFA9626378E412B443CFE2C4BC2DF5C4649F3303AFAC38435) 
I[2020-10-21|16:33:41.234] - [stargate-4]@{94581}conn(adityasrcconn)-{STATE_INIT} : [slippy-0]@{102975}conn(adityadstconn)-{STATE_TRYOPEN} 
I[2020-10-21|16:33:41.844] - [stargate-4] -> sending transaction:       
E[2020-10-21|16:33:41.844] - [stargate-4] -> err(failed to execute message; message index: 1: connection handshake open ack failed: connection state is not INIT or TRYOPEN (got STATE_OPEN): invalid connection state: invalid request) 
^C
➜  relayer git:(aditya/testing) ✗ rly tx full-path demo -d -o 3s
I[2020-10-21|16:33:51.605] - [slippy-0]@{102977}conn(adityadstconn)-{STATE_TRYOPEN} : [stargate-4]@{94582}conn(adityasrcconn)-{STATE_OPEN} 
I[2020-10-21|16:33:54.349] ✔ [slippy-0]@{102978} - msg(0:update_client,1:connection_open_confirm) hash(E1AA601A0DF5B436D2DCE4A6FCF86A99FF778E9E85B2946F6F5E05DAEB311F21) 
I[2020-10-21|16:33:54.661] - [stargate-4]@{0}conn(adityasrcconn)-{STATE_UNINITIALIZED_UNSPECIFIED} : [slippy-0]@{0}conn(adityadstconn)-{STATE_UNINITIALIZED_UNSPECIFIED} 
I[2020-10-21|16:33:54.661] ★ Connection created: [stargate-4]client{adityasrcclient}conn{adityasrcconn} -> [slippy-0]client{adityadstclient}conn{adityadstconn} 
I[2020-10-21|16:33:55.655] - [stargate-4]@{0}chan(adityasrcchan)-{STATE_UNINITIALIZED_UNSPECIFIED} : [slippy-0]@{0}chan(adityadstchan)-{STATE_UNINITIALIZED_UNSPECIFIED} 
I[2020-10-21|16:33:57.741] ✔ [stargate-4]@{94584} - msg(0:channel_open_init) hash(9759CFB596A169B25A24321656ABBA8766305CA66536AE9A67865DED2086E63D) 
I[2020-10-21|16:33:58.737] - [stargate-4]@{0}chan(adityasrcchan)-{STATE_UNINITIALIZED_UNSPECIFIED} : [slippy-0]@{0}chan(adityadstchan)-{STATE_UNINITIALIZED_UNSPECIFIED} 
I[2020-10-21|16:33:59.330] - [stargate-4] -> sending transaction:       
E[2020-10-21|16:33:59.331] - [stargate-4] -> err(failed to execute message; message index: 0: channel handshake open init failed: port ID (transfer) channel ID (adityasrcchan): channel already exists: invalid request) 
^C
➜  relayer git:(aditya/testing) ✗ 
➜  relayer git:(aditya/testing) ✗ rly tx full-path demo -d -o 3s
I[2020-10-21|16:35:34.502] - [slippy-0]@{0}chan(adityadstchan)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{94599}chan(adityasrcchan)-{STATE_INIT} 
I[2020-10-21|16:35:39.561] ✔ [slippy-0]@{102999} - msg(0:update_client,1:channel_open_try) hash(F8B9EAE1218ABF4962A8F69DD99E7AD0E89C293FC48846CEC40D495769D439C7) 
I[2020-10-21|16:35:41.117] - [slippy-0]@{0}chan(adityadstchan)-{STATE_UNINITIALIZED_UNSPECIFIED} : [stargate-4]@{94601}chan(adityasrcchan)-{STATE_INIT} 
I[2020-10-21|16:35:41.595] - [slippy-0] -> sending transaction:         
E[2020-10-21|16:35:41.596] - [slippy-0] -> err(failed to execute message; message index: 1: channel handshake open try failed: cannot relay connection attempt: invalid channel: invalid request) 
^C
➜  relayer git:(aditya/testing) ✗ rly tx full-path demo -d -o 3s
I[2020-10-21|16:35:58.090] - [stargate-4]@{94604}chan(adityasrcchan)-{STATE_INIT} : [slippy-0]@{103002}chan(adityadstchan)-{STATE_TRYOPEN} 
I[2020-10-21|16:36:02.589] ✔ [stargate-4]@{94605} - msg(0:update_client,1:channel_open_ack) hash(CC4C18EE31C7256FB789BB26EFF08A337F4E34EDB3CFFD517407CE27A8B2624E) 
I[2020-10-21|16:36:03.740] - [stargate-4]@{94605}chan(adityasrcchan)-{STATE_INIT} : [slippy-0]@{103003}chan(adityadstchan)-{STATE_TRYOPEN} 
I[2020-10-21|16:36:04.642] - [stargate-4] -> sending transaction:       
E[2020-10-21|16:36:04.642] - [stargate-4] -> err(failed to execute message; message index: 1: channel handshake open ack failed: channel state should be INIT or TRYOPEN (got STATE_OPEN): invalid channel state: invalid request) 
^C
➜  relayer git:(aditya/testing) ✗ rly tx full-path demo -d -o 3s
I[2020-10-21|16:36:14.109] - [slippy-0]@{103005}chan(adityadstchan)-{STATE_TRYOPEN} : [stargate-4]@{94606}chan(adityasrcchan)-{STATE_OPEN} 
I[2020-10-21|16:36:19.610] ✔ [slippy-0]@{103007} - msg(0:update_client,1:channel_open_confirm) hash(FE78333ECDB772512F261C43F4EF85E73B5209CA3CEF4283D12CB4CDAE846658) 
I[2020-10-21|16:36:19.937] - [stargate-4]@{0}chan(adityasrcchan)-{STATE_UNINITIALIZED_UNSPECIFIED} : [slippy-0]@{0}chan(adityadstchan)-{STATE_UNINITIALIZED_UNSPECIFIED} 
I[2020-10-21|16:36:19.937] ★ Channel created: [stargate-4]chan{adityasrcchan}port{transfer} -> [slippy-0]chan{adityadstchan}port{transfer} 
```

Simply rerunning the command over and over again allows the handshake to progress

**Comment by jackzampolin** [comment]
Pretty sure now that the locks are fixed this issue is no longer present.

**Comment by colin-axner** [comment]
this no longer occurs with fixes in #334
